webextension/background.js

```diff
@@ -129,8 +127,75 @@ const Config = (function() {
       }
     }
 
-    load() {
+    maybeSendTelemetry(message) {
+      browser.study.sendTelemetry(message);
+    }
+
+    get shieldStudySetup() {
+      const pref = `extensions.${browser.runtime.id.split("@")[0]}.variation`;
+      return {
+        activeExperimentName: browser.runtime.id,
+        allowEnroll: true,
+        studyType: "shield",
+        telemetry: {
+          send: true,
+          removeTestingFlag: !this._testingMode,
+        },
+        endings: {
+          "user-disable": {
+            baseUrl: null,
+          },
+          expired: {
+            baseUrl: null,
+          },
+          dataPermissionsRevoked: {
+            baseUrl: null,
+            study_state: "ended-neutral",
+          },
+        },
+        logLevel: this._testingMode ? 30 : 0,
+        variationOverridePreference: pref,
+        weightedVariations: UIVariants.map(name => {
+          return {name, weight: 1};
+        }),
+        expire: {
+          days: 14,
+        },
+      };
+    }
+
+    _activateShield() {
+      if (this._shieldActivatedPromise) {
+        return this._shieldActivatedPromise;
+      }
+
+      this._shieldActivatedPromise = new Promise((resolve, reject) => {
+        const endListener = studyInfoOrError => {
```

Comment:
I believe that one of the magical aspects of web extension `onWhatever` listeners is that you don't need to manage all this teardown.   Internally, it has a WeakMap that tracks it tearsdown during `callOnClose`. 

What you do here is Safe, but verbose and potentially easy to create a bug.  Confirm it with irc teamAddons or friends though!

Reply:
Yes, I just wasn't 100% sure whether either method will be called more than once as the addon runs, so I wanted to guard against that it triggering a now-stale listener and messing up internal state.

I'm not worried about this, but don't mind changing it if anyone has any strong opinions on simplifying it (since this addon may not be a Shield-driven one forever anyhow).